File: typescript/examples/sign_message.ts
import { TronProvider } from '../src/wallet';

function usage(): never {
  // eslint-disable-next-line no-console
  console.error(
    [
      'Usage:',
      '  LOG_LEVEL=info TRON_PRIVATE_KEY=... node -r ts-node/register examples/sign_message.ts "hello" [utf8|hex]',
      '',
      'Notes:',
      '  - If TRON_PRIVATE_KEY is not set, TronProvider will try to load privateKey from keystore (~/.agent_wallet/Keystore).',
      '  - For encoding=hex, the message must be a hex string (no 0x prefix).',
    ].join('\n')
  );
  process.exit(2);
}

async function main(): Promise<void> {
  const message = process.argv[2];
  const encoding = (process.argv[3] ?? 'utf8') as 'utf8' | 'hex';
  if (!message) usage();
  if (encoding !== 'utf8' && encoding !== 'hex') usage();

  const provider = new TronProvider({
    // All of these are optional; provider can load from env + keystore
    privateKey: process.env.TRON_PRIVATE_KEY,
    apiKey: process.env.TRON_GRID_API_KEY,
    keystore: { password: process.env.KEYSTORE_PASSWORD },
  });
  await provider.init(); // loads credentials from keystore if needed

  const res = await provider.signTx({ type: 'message', message, encoding });
  // eslint-disable-next-line no-console
  console.log(res.signature ?? '');
}

main().catch((e) => {
  // eslint-disable-next-line no-console
  console.error(String(e?.message ?? e));
  process.exit(1);
});

File: python/examples/sign_message.py
import os
import sys
import asyncio

from pathlib import Path

# Ensure `python/` directory is on sys.path when running as a script:
# `python examples/sign_message.py ...` sets sys.path[0] to `python/examples`.
sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from wallet import TronProvider


def usage() -> None:
    print(
        "\n".join(
            [
                "Usage:",
                '  LOG_LEVEL=INFO TRON_PRIVATE_KEY=... .venv/bin/python examples/sign_message.py "hello" [utf8|hex]',
                "",
                "Notes:",
                "  - If TRON_PRIVATE_KEY is not set, TronProvider will try to load privateKey from keystore (~/.agent_wallet/Keystore).",
                "  - For encoding=hex, the message must be a hex string (no 0x prefix).",
            ]
        )
    )


async def main() -> int:
    if len(sys.argv) < 2:
        usage()
        return 2

    message = sys.argv[1]
    parse_as = sys.argv[2] if len(sys.argv) >= 3 else "utf8"
    if parse_as not in ("utf8", "hex"):
        usage()
        return 2

    message_bytes = bytes.fromhex(message) if parse_as == "hex" else message.encode("utf-8")

    provider = TronProvider(
        private_key=os.getenv("TRON_PRIVATE_KEY"),
        api_key=os.getenv("TRON_GRID_API_KEY"),
        keystore_path=os.getenv("KEYSTORE_PATH"),
    )
    await provider.init()

    sig = await provider.sign_message(message_bytes)
    print(sig)
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

